OpenCV 3.2.0 won't see ffmpeg on cmake. Versions 3.1.0 and 3.0.0 will

##### System information (version)

- OpenCV => 3.2.0
- Operating System / Platform => Debian Linux 7.11 (Wheezy image) / Beaglebone Black ARM 32 bits 
- Compiler => GCC 4.6.3

##### Detailed description

- OpenCV 3.2.0 won't see ffmpeg on cmake setup even with ffmpeg properly set and the -D WITH_FFMPEG=ON flag.

After OpenCV setup (without identifying the machine's ffmpeg library) I can properly build programs without compiler errors but whenever I try to open a h264 video set I get the following runtime error:

"Unable to stop the stream: Inappropriate ioctl for device"

- Both OpenCV 3.1.0 and OpenCV 3.0.0 will see ffmpeg and properly setup and compile the library in the exact same setup, and programs using them will open and process h264 videos properly.

##### Steps to reproduce

- Install ffmpeg from debian repository or compile it yourself. You will get the error both ways.

- Configure OpenCV 3.2.0 with the following cmake setup:

cmake -D WITH_OPENCL=OFF -D BUILD_TESTS=OFF -D BUILD_PERF_TESTS=OFF -D WITH_FFMPEG=ON ..

You will get the following setup output:
...
--   Video I/O:
--     DC1394 1.x:                  NO
--     DC1394 2.x:                  YES (ver 2.2.0)
--     FFMPEG:                      NO
--       codec:                     YES (ver 57.75.100)
--       format:                    YES (ver 57.65.100)
--       util:                      YES (ver 55.45.100)
--       swscale:                   YES (ver 4.3.101)
--       resample:                  NO
--       gentoo-style:              YES
...

- Compile and install it as usual:

make
sudo make install

- Create a c++ program with the following lines:

#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>

int main (int argc, char **argv) {
   cv::VideoCapture video;
   cv::namedWindow("Original", cv::WINDOW_AUTOSIZE | CV_GUI_NORMAL);
   video.open("your_video_H264.mkv");
   video.read(frame);
   cv::imshow("Original", frame);
   cv::waitKey(0);
   video.release();
   return 1;
}

- Try to open a h264 video embedded in a .mkv matroska container and it will throw the following error:

"Unable to stop the stream: Inappropriate ioctl for device"

Got the same issue here, noticed this while running cmake : 
`WARNING: Can't build ffmpeg test code`

So I uncommented cmake/OpenCVFindLibsVideo.cmake:238 : 
`message(FATAL_ERROR "FFMPEG: test check build log:\n${TRY_OUT}")`

It prints a whole lot of link error in avutil.a : 
```
FFmpeg/libavutil/hwcontext_vaapi.c:718: undefined reference to `vaSyncSurface'
...
```

I think it's related to the link order while compiling the test program against statically build avutil. Don't known much about pkg-config files but the one generated by ffmpeg for avutil seems wrong so I replaced the `Libs` line by : 
```
Libs: -L${libdir}  -lavutil -lm -lvdpau -lva -lva-drm -lva-x11 -lX11
```

Just cleaned the build folder and reinvoking cmake correctly find FFMPEG : 
```
--     FFMPEG:                      YES
--       avcodec:                   YES (ver 57.83.100)
--       avformat:                  YES (ver 57.66.104)
--       avutil:                    YES (ver 55.48.100)
--       swscale:                   YES (ver 4.3.101)
--       avresample:                NO
```

--- a/cmake/checks/ffmpeg_test.cpp
+++ b/cmake/checks/ffmpeg_test.cpp
@@ -1,3 +1,5 @@
+#define __STDC_CONSTANT_MACROS
+
 #include <stdlib.h>
 
 extern "C" {

